src/linter.ts

```diff
@@ -131,19 +131,40 @@ class Linter {
             throw new Error(`formatter '${formatterName}' not found`);
         }
 
-        const output = formatter.format(this.failures, this.fixes);
+        // Sort failures by failure's line and character.
+        const failures = this.sortFailures(this.failures);
+        const fixes = this.sortFailures(this.fixes);
+        const output = formatter.format(failures, fixes);
 
         const errorCount = this.failures.filter((failure) => failure.getRuleSeverity() === "error").length;
         return {
             errorCount,
-            failures: this.failures,
-            fixes: this.fixes,
+            failures,
+            fixes,
             format: formatterName,
             output,
             warningCount: this.failures.length - errorCount,
         };
     }
 
+    public sortFailures(failures: RuleFailure[]): RuleFailure[] {
+        let sortedFailures: RuleFailure[] = [];
+
+        const fileFailuresMap = createMultiMap(failures, (failure) => [failure.getFileName(), failure]);
+        fileFailuresMap.forEach((fileFailures) => {
+            sortedFailures = sortedFailures.concat(fileFailures.sort((failureA, failureB) => {
+                const failureALineAndCharacter = failureA.getStartPosition().getLineAndCharacter();
+                const failureBLineAndCharacter = failureB.getStartPosition().getLineAndCharacter();
+                if (failureALineAndCharacter.line === failureBLineAndCharacter.line) {
+                    return failureALineAndCharacter.character - failureBLineAndCharacter.character;
+                }
+                return failureALineAndCharacter.line - failureBLineAndCharacter.line;
+            }));
+        });
+
+        return sortedFailures;
+    }
```

Comment:
As described above, I'd like to move this to the formatters. To make the compare logic reusable, consider adding it as static method to `class RuleFailure` in `src/language/rule/rule.ts`:

```ts
    public static compare(a: RuleFailure, b: RuleFailure): number {
        if (a.fileName !== b.fileName) {
            return a.fileName < b.fileName ? -1 : 1;
        }
        return a.startPosition.getPosition() - b.startPosition.getPosition();
    }
```